Assets/Scripts/Entity/Enemy/DetectPlayer.cs

```diff
@@ -0,0 +1,80 @@
+﻿using System.Collections;
+using System.Collections.Generic;
+using UnityEngine;
+
+
+[RequireComponent(typeof(FieldOfView))]
+public class DetectPlayer : MonoBehaviour
+{
+    private FieldOfView fov;
+
+    [Tooltip("Behavior scripts, if any, that should be disabled when the player is detected")]
+    [SerializeField] private MonoBehaviour[] behaviorsToDisableUponDetection;
+
+    [Tooltip("Behavior scripts, if any, that should be enabled when the player is detected")]
+    [SerializeField] private MonoBehaviour[] behaviorsToEnableUponDetection;
+
+
+    /* Called before the first frame update.
+     */
+    private void Start()
+    {
+        fov = GetComponent<FieldOfView>();
+        GetComponent<Enemy>().healthChangedEvent += OnEnemyHurt;
+        StartCoroutine(FindPlayerWithDelay(0.2f));
+    }
+
+
+    /* Called when the enemy is hit by a player projectile. This is treated
+     * as a detection.
+     */
+    private void OnEnemyHurt()
+    {
+        DisableBehaviors();
+        EnableBehaviors();
+        fov.viewRadius = 40;
+        enabled = false;
+    }
+
+
+    /* Disables all pre-player-detection behaviors.
+     */
+    private void DisableBehaviors()
+    {
+        foreach (MonoBehaviour script in behaviorsToDisableUponDetection)
+        {
+            script.enabled = false;
+        }
+    }
+
+
+    /* Enables all post-player-detection behaviors.
+     */
+    private void EnableBehaviors()
+    {
+        foreach (MonoBehaviour script in behaviorsToEnableUponDetection)
+        {
+            script.enabled = true;
+        }
+    }
+
+
+    /* Looks for the player character with the given delay in seconds
+     * between each search.
+     */ 
+    private IEnumerator FindPlayerWithDelay(float delay)
+    {
+        while (enabled)
+        {
+            if(fov.PlayerWithinView())
+            {
+                DisableBehaviors();
```

Comment:
hmmm I have a couple questions about these 4 lines of code.

1. Why is it duplicated above? (Im not sure if they are meant to be identical or it's just a coincidence)
2. Why is this disabling and then enabling the behaviors? Is there a specific reason to do that.
3. Seems like the viewRadius is being overriden here, regardless of what it may have been set in the FieldOfVision script.
4. The script is continuously disabled here but (a) the script is never enabled, and (b) disabling this script won't do anything (I dont think) because it has no Update method.

Reply:
You need to get more comfortable with reading code; most of the stuff I did here is documented with comments, and the rest is self-explanatory because the variable names are purposeful.

1. One handles player detection when the enemy is shot at (`OnEnemyHurt`). This one handles player detection when the enemy actually _spots_ the player. They use the same lines of code, just in different scenarios.

2. It's disabling and enabling behaviors to facilitate a state transition, from a state prior to when the player was detected to one after the player was detected. If you look at the enemy prefabs, you'll find that all of them use a pre-detection behavior of `WanderMovement` and a post-detection behavior of `AIAttackController` and `PursuePlayer`. In other words, `DetectPlayer` allows an enemy to stop doing one set of behaviors and start doing another set of behaviors once it has detected the player.

3. Yes, it's being overriden because the player has already been detected. The idea is to make the view radius quite large so that when the enemies are aiming with their field of view (see `AIAttackController`), the player is always within the radius but perhaps not within the _cone_. In other words, it's to give all enemies an "infinite" shooting distance/radius within the room.

4. How is the script continuously disabled here? The co-routine stops as soon as the script is disabled.

Reply:
Yeah no, it's not totally obvious that these 4 lines are a state transition. `DiableBehaviors` and `EnableBehaviors` are not very descriptive and neither is a `viewRadius` of `40` (Magic number). It's not obvious that these 4 lines are causing a state change from that alone, they could have been in a `DetectedPlayerStateTransition` method. Also, concerning the disable, [Note: Coroutines are not stopped when a MonoBehaviour is disabled](https://docs.unity3d.com/Manual/Coroutines.html), which only made this even more confusing since it was not clear that the coroutine was NOT supposed to be continuously running, disabling and enabling behaviors.

> They use the same lines of code, just in different scenarios.

Could be packaged in a nice method. That way if we ever change this "state transition" we won't accidentally miss one or the other.

Reply:
Your concerns are fair (though `fov` is a standard abbreviation for field of view in gaming).

Re: the coroutine not stopping, it only runs while the script is enabled (`while(enabled)`). So as soon as the script gets disabled, the coroutine will return.